### ISSUE TITLE
Is per_device_loader mutating the output of the DataLoader?

### ISSUE THREAD

**Issue by hrbigelow** [pytorch/xla#1451]
## 🐛 Bug

UPDATE:  I have seen other bizarre things happening which are not related to ParallelLoader.  I'm wondering if there is some wider issue causing it that is not specific to pytorch_xla.

I am wrapping my `torch.utils.data.DataLoader` in a ParallelLoader per_device_loader.  The DataLoader object's `__next__` method returns a VirtualBatch, which has several CPU tensors of different shapes.  However, when per_device_loader is called on it, it returns a VirtualBatch alright, but the shapes of the tensors are different.  Even more weird, when I debug, and inspect the output of the per_device_loader (confirming it is wrong), I can interactively repeat the command, and get an object that has the right shape!

In the example below, I inspect the VirtualBatch that was returned.  The shapes are wrong.  I then 
call the next method again, and the returned `vb2` has the correct shapes.

However, this is done using my own build of pytorch (ea3697db) and pytorch-xla (e49ac485).

I will test this on nightly once the next build of nightly succeeds.

```
/content/ae-wavenet/model.py(234)__next__()
    232     def __next__(self):
1   233         vb = self.per_dev_loader.__next__()[0]
--> 234         assert vb.wav_dec_input.shape[0] == 8
    235         return vb
    236 

ipdb> vb
voice_index: tensor([122, 122, 122,  ..., 126, 127, 126], device='xla:1', dtype=torch.int32)
jitter_index: tensor([2235, 4381], device='xla:1')
wav_dec_input.shape: torch.Size([2])
mel_enc_input.shape: torch.Size([2])

ipdb> vb2 = self.per_dev_loader.__next__()[0]
ipdb> vb2
voice_index: tensor([8, 5, 6, 8, 2, 0, 4, 6], device='xla:1')
jitter_index: tensor([[  0,   1,   2,   3,   3,   5,   6,   6,   8,   9,  11,  11,  12,  13],
        [ 14,  15,  16,  18,  18,  19,  21,  21,  22,  23,  23,  25,  26,  27],
        [ 28,  29,  31,  31,  32,  33,  34,  35,  36,  37,  38,  39,  41,  41],
        [ 42,  43,  43,  45,  46,  48,  48,  50,  50,  51,  52,  53,  54,  55],
        [ 56,  57,  58,  59,  60,  60,  62,  63,  64,  65,  65,  66,  68,  69],
        [ 70,  71,  72,  73,  74,  74,  76,  77,  78,  79,  80,  81,  82,  83],
        [ 84,  85,  87,  87,  88,  89,  90,  92,  91,  92,  93,  95,  96,  97],
        [ 98,  99, 101, 101, 102, 102, 104, 105, 106, 107, 108, 109, 110, 111]],
       device='xla:1')
wav_dec_input.shape: torch.Size([8, 2146])
mel_enc_input.shape: torch.Size([8, 39, 42])
```

## To Reproduce

1.  Go to this [Colab](https://colab.research.google.com/drive/1MDY5ZpjlIU9Np4r0xSFNevtGekZT6vZk)
2.  Run using my wheels:  [xla](https://drive.google.com/open?id=1HnMLFkAdm1-r3gDfujvPERz1BJ6_CyIN) and [torch](https://drive.google.com/open?id=1pGgNgbgCpZdkJiC-NnVzpoxnU_u02Hyp)
3.  Use the commands:

```
# !cd ae-wavenet && git pull
!cd ae-wavenet && git checkout master
!cd ae-wavenet && git pull
!cd ae-wavenet && git checkout 0a5d6b1

# delete all currently loaded ae-wavenet modules
import sys
d = list(map(lambda m: m[0], filter(lambda m: hasattr(m[1], '__file__') and 'ae-wavenet' in m[1].__file__, sys.modules.items())))
def dn(n): del sys.modules[n]
list(map(dn, d))

# %env XLA_USE_32BIT_LONG=0
# %env XLA_DENSE_GATHER_FACTOR=1
VER=!git -C ae-wavenet rev-parse --short HEAD
!echo ae-wavenet commit used: {VER}
%run -b ae-wavenet/model.py:233 -d ae-wavenet/train.py new -af ae-wavenet/par/arch.ae.json -tf ae-wavenet/par/train.basic.json -hw TPU-single -pi 1 -dlo 8 -dng 0 -lrr 1e-6 -lrs 0 -nw 100 -nb 8 -si 5000 ./basic.full.%.ckpt ae-wavenet/dat/librispeech.some.dat
# !./xla/scripts/debug_run.py --tidy --outfile ./gdrive/My\ Drive/ai/{VER[0]}.dlo128.debug_run.tar.gz -- python ae-wavenet/train.py new -ms 20 -dlo 128 -pi 20 -af ae-wavenet/par/arch.ae.json -tf ae-wavenet/par/train.basic.json -hw TPU-single -dng 0 -lrr 1e-6 -lrs 0 -nw 100 -nb 8 -si 5000 ./basic.full.%.ckpt ae-wavenet/dat/librispeech.some.dat
# !python
```

## Environment

 - reproducible on XLA backend [CPU/TPU]:
 - torch_xla version:
 - Any other relevant information. In case of issues related to slow executions, re-compilations, etc.., it is useful to have the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

**Comment by dlibenzi** [comment]
Can you please create a smaller repro, as likely all the carryover from the main model is unrelated?
Just have you iterator return a couple of your virtual batch objects, pass your data-loader to the parallel_loader, and compare output shapes.

**Comment by hrbigelow** [comment]
Sounds good.  I tried it with nightly and it exhibits the same two symptoms.  I'll create a minimal example and report back

**Comment by hrbigelow** [comment]
Okay, so, I found out what is happening.  It's a subtle kind of bug, and likely an edge case.  It has to do with the fact that I'd implemented `__getstate__` and `__setstate__` for my `DataLoader`.  However, I use these to save/restore from checkpoint.  But, I also call a function that does further initialization *after* restoring from checkpoint.  I believe that ParallelLoader clones the DataLoader using `__getstate__` and `__setstate__`, but at one point, it is dealing with both a version with the post initialization, and a version of the instance without.  I'll post a minimal example shortly.

**Comment by hrbigelow** [comment]
@dlibenzi Okay, I have a minimal example, see [data_loader_test.py](https://github.com/hrbigelow/xla-tests/blob/master/data_loader_test.py) or try it out in this [Colab](https://colab.research.google.com/drive/1NtaPpsZmrQ02qJhBeSyS4La32sg25tjS)

The reason it is convenient for me to use `__getstate__` and `__setstate__` in the way described above is that my data and model modules have a bit of messy dependence on each other, and so I can't restore them entirely in one step.  However, I'm not sure whether it is breaking a "contract" of the intended use of those functions, is it?

**Comment by dlibenzi** [comment]
Your problem is that your Batch data structure has pointers back to the Dataset.
Simple solution, detach the real batch (which should only contain data) from the parent Dataset/DataLoader:

https://gist.github.com/dlibenzi/224e8441b6e784f62b37f0b2dc8dd66e

**Comment by hrbigelow** [comment]
Got it, thanks!  I made that change and the batches all have the same shape now.
